Mathlib/Computability/CFG.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar (CFG), which is a grammar that has a
+single nonterminal symbol on the left-hand side of each rule.
+Note that derivation by a grammar is inherently nondeterministic.
+-/
+
+/-- The type of symbols is the disjoint union of terminals and nonterminals. -/
+inductive Symbol (T : Type) (N : Type)
+  | terminal    (t : T) : Symbol T N
+  | nonterminal (n : N) : Symbol T N
+/-
+We do not require T and N to be finite.
+As a result, we do not need to copy the typeclass instances `[Fintype T]` and `[Fintype N]`
+alongside our type parameters (which would appear in almost every lemma).
+Instead, later we work in terms of a list of rewrite rules, which is finite by definition and from
+which we could infer that only a finite set of terminals and a finite set of nonterminals can occur.
+-/
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure CFG (T : Type) where
+  nt : Type                              -- type of nonterminals
+  initial : nt                           -- initial nonterminal
+  rules : List (nt × List (Symbol T nt)) -- rewrite rules
+
+variable {T : Type}
```

Comment:
probably you should open a CFG namespace here

Reply:
Do you want `open CFG` or `namespace CFG` here? I believe the former accomplishes nothing because I use the `.` notation everywhere.

Reply:
It accomplishes not having to write `CFG.` everywhere, which imo helps readability

Reply:
Then I have to write `namespace CFG` right? The prefix `CFG.` is written only in the declarations; calls never use it explicitly.

Reply:
I don't want `IsCF` to become `CFG.IsCF` because it should be `Language.IsCF` by semantics. What should I do?

Reply:
end the namespace before declaring `IsCF` (or alternatively use `_root_.Language.IsCF`
You should put the new lemmas above `IsCF` in that case, next to the defs they are lemmas about.

Reply:
Is it worth opening the namespace then? Since I use the dot notation everywhere, very little will be saved by `CFG.` not being repeated.

Reply:
OK, I'll put it into a namespace then. Just tell me, should I use `g.Derives u v` or `Derives g u v` now? My switch to the former notation was motivated by not having `CFG.` prepended automatically.